File: src/main/java/org/jenkinsci/gradle/plugins/jpi/GitVersionExtension.java
package org.jenkinsci.gradle.plugins.jpi;

import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.ProjectLayout;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.ProviderFactory;
import org.gradle.api.tasks.Optional;

import javax.inject.Inject;

/**
 * Gradle extension for configuring Git version generation.
 * <p>
 * This extension provides configuration options for generating version strings
 * based on Git repository information.
 */
public abstract class GitVersionExtension {

    private static final int DEFAULT_ABBREV_LENGTH = 12;
    private static final String DEFAULT_VERSION_FORMAT = "%d.%s";

    /**
     * Constructs a new Git version extension with default values.
     *
     * @param layout The project layout for accessing project directories
     * @param providers The provider factory for accessing Gradle properties
     */
    @Inject
    public GitVersionExtension(ProjectLayout layout, ProviderFactory providers) {
        getVersionFormat().convention(
                providers.gradleProperty("gitVersionFormat").orElse(DEFAULT_VERSION_FORMAT));
        getAbbrevLength().convention(DEFAULT_ABBREV_LENGTH);
        getAllowDirty().convention(false);
        getGitRoot().convention(layout.getProjectDirectory());
        getSanitize().convention(providers.gradleProperty("gitVersionSanitize").map(p -> true).orElse(false));
        getOutputFile().convention(
                providers.gradleProperty("gitVersionFile")
                        .map(p-> layout.getProjectDirectory().file(p))
                        .orElse(layout.getBuildDirectory().file("generated/version/version.txt")));
        getVersionPrefix().convention("");
    }

    /**
     * Gets the format string for the version.
     * <p>
     * The format string is used with String.format() and receives the commit depth
     * and abbreviated hash as arguments.
     *
     * @return A property containing the version format
     */
    @Optional
    public abstract Property<String> getVersionFormat();

    /**
     * Gets the prefix to use for the version string.
     *
     * @return A property containing the version prefix
     */
    @Optional
    public abstract Property<String> getVersionPrefix();

    /**
     * Gets whether to sanitize the version string.
     * <p>
     * When true, certain characters in the Git hash are replaced to make the
     * version string more compatible with various systems.
     *
     * @return A property indicating whether to sanitize the version
     */
    @Optional
    public abstract Property<Boolean> getSanitize();

    /**
     * Gets the abbreviation length for commit hashes.
     *
     * @return A property containing the abbreviation length
     */
    @Optional
    public abstract Property<Integer> getAbbrevLength();

    /**
     * Gets whether to allow dirty working directory.
     * <p>
     * When false, the version generation will fail if the Git repository
     * has uncommitted changes.
     *
     * @return A property indicating whether to allow dirty working directory
     */
    @Optional
    public abstract Property<Boolean> getAllowDirty();

    /**
     * Gets the root directory of the Git repository.
     *
     * @return A property containing the Git root directory
     */
    @Optional
    public abstract DirectoryProperty getGitRoot();

    /**
     * Gets the output file where the generated version will be written.
     *
     * @return A property containing the output file location
     */
    @Optional
    public abstract RegularFileProperty getOutputFile();


}

File: jpi2/src/main/kotlin/org/jenkinsci/gradle/plugins/jpi2/GitVersionExtension.kt
package org.jenkinsci.gradle.plugins.jpi2

import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.ProjectLayout
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.provider.Provider
import org.gradle.api.provider.ProviderFactory
import javax.inject.Inject

/**
 * Configuration for Git-based version generation (used when [JenkinsPluginExtension.versionSource] is [VersionSource.GIT]).
 *
 * The version string is computed from the Git repository: commit depth and abbreviated hash,
 * formatted with [versionFormat] (default `%d.%s` for depth.hash). Written to [outputFile];
 * the first line is the version string used for the plugin.
 */
abstract class GitVersionExtension @Inject constructor(
    objectFactory: ObjectFactory,
    private val layout: ProjectLayout,
    providers: ProviderFactory,
) {

    companion object {
        const val DEFAULT_VERSION_FORMAT = "%d.%s"
        const val DEFAULT_ABBREV_LENGTH = 12
    }

    /**
     * Format string for the version. Receives commit depth (Long) and abbreviated hash (String).
     * Default is [DEFAULT_VERSION_FORMAT] (`%d.%s`). Can be overridden with Gradle property `gitVersionFormat`.
     */
    val versionFormat: Property<String> = objectFactory.property(String::class.java)
        .convention(providers.gradleProperty("gitVersionFormat").orElse(DEFAULT_VERSION_FORMAT))

    /**
     * Prefix prepended to the formatted version string. Can be set via Gradle property `gitVersionPrefix`.
     */
    val versionPrefix: Property<String> = objectFactory.property(String::class.java)
        .convention(providers.gradleProperty("gitVersionPrefix").orElse(""))

    /**
     * Length of the abbreviated Git hash. Default [DEFAULT_ABBREV_LENGTH].
     */
    val abbrevLength: Property<Int> = objectFactory.property(Int::class.javaObjectType)
        .convention(DEFAULT_ABBREV_LENGTH)

    /**
     * Whether to allow a dirty working directory. If false, the task fails when there are uncommitted changes.
     */
    val allowDirty: Property<Boolean> = objectFactory.property(Boolean::class.javaObjectType)
        .convention(false)

    /**
     * Root of the Git repository. Defaults to the project directory.
     */
    val gitRoot: DirectoryProperty = objectFactory.directoryProperty()
        .convention(layout.projectDirectory)

    /**
     * Output file for the generated version (first line = version string, second line = full hash).
     * Can be set via Gradle property `gitVersionFile`. Defaults to `build/generated/version/version.txt`.
     */
    val outputFile: RegularFileProperty = objectFactory.fileProperty()
        .convention(
            providers.gradleProperty("gitVersionFile")
                .map { layout.projectDirectory.file(it) }
                .orElse(layout.buildDirectory.file("generated/version/version.txt"))
        )

    /**
     * Provider for the Git-derived version string (commit depth + abbreviated hash per [versionFormat]).
     * Resolved on demand when the value is read; no task required. Use this as the source of truth
     * for the plugin version when [VersionSource.GIT]. The [generateGitVersion] task still writes
     * this value (and the full hash) to [outputFile] for scripts and compatibility.
     */
    val version: Provider<String> = providers.provider {
        GitVersion.compute(
            gitRoot.get().asFile.toPath(),
            versionFormat.get(),
            versionPrefix.get(),
            abbrevLength.get(),
            allowDirty.get(),
        ).version()
    }
}
